[TensorPipe] Do not require user to provide worker name-to-rank map

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38052 [TensorPipe] Do not require user to provide worker name-to-rank map**

The initial version of the TensorPipe agent required the user to specify the full map between workers' names and their ids, on each worker. However it's enough for each worker to just specify their name and id, as these can then be exchanged using the store.

Addresses #37784, although I think we can go further and use the store to also automatically assign ranks to workers, so that the user only needs to specify a name.

Differential Revision: [D21463833](https://our.internmc.facebook.com/intern/diff/D21463833/)

lw committed:
[TensorPipe] Do not require user to provide worker name-to-rank map

The initial version of the TensorPipe agent required the user to specify the full map between workers' names and their ids, on each worker. However it's enough for each worker to just specify their name and id, as these can then be exchanged using the store.

Addresses #37784, although I think we can go further and use the store to also automatically assign ranks to workers, so that the user only needs to specify a name.

Differential Revision: [D21463833](https://our.internmc.facebook.com/intern/diff/D21463833/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -19,10 +19,36 @@ namespace rpc {
+  std::vector<uint8_t> selfNameVector(
+      (uint8_t*)selfName.c_str(),
+      (uint8_t*)selfName.c_str() + selfName.length());
+  addressStore_->set("names/" + std::to_string(selfId), selfNameVector);
```
nit: There is a convention to use `c10::to_string` in our code base.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -19,10 +19,36 @@ namespace rpc {
+    }
+    std::string workerKey = "names/" + std::to_string(workerId);
+    addressStore_->wait({workerKey});
+    std::vector<uint8_t> workerNameVector = addressStore_->get(workerKey);
```
Seems `get()` is implicitly doing `wait()`?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -32,9 +26,10 @@ struct TensorPipeRpcBackendOptions : public RpcBackendOptions {
   TensorPipeAgent(
-      worker_id_t selfId,
-      std::string selfName,
       std::shared_ptr<::c10d::Store> addressStore,
```
cc @kiukchung @tierex `TensorPipeAgent` is already taking a c10d::Store

mrshenli reviewed (approved):
Thanks for adding this. Test failures do not look relevant, but if master is getting better early tomorrow. Could you please rebase to make sure we don't break anything?

Another thing is that, test coverage becomes more and more urgent. I am timid to keep adding code with no test coverage. It will be great to see TensorPipe getting connected to (subset of) our existing RPC tests.

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -19,10 +19,36 @@ namespace rpc {
     std::shared_ptr<::c10d::Store> addressStore,
+    std::string selfName,
+    worker_id_t selfId,
+    worker_id_t worldSize,
```
I think we use `int` for `worldSize` instead of `worker_id_t`? Looks like worker_id_t is used only for ID's.

osalpekar commented:
> Thanks for adding this. Test failures do not look relevant, but if master is getting better early tomorrow. Could you please rebase to make sure we don't break anything?
> 
> Another thing is that, test coverage becomes more and more urgent. I am timid to keep adding code with no test coverage. It will be great to see TensorPipe getting connected to (subset of) our existing RPC tests.

@mrshenli I have a diff enabling the RPC, Dist Autograd, and Dist Optimizer tests with the TensorPipe Agent (D21412016). Merging this PR will make the tests run more smoothly in the current framework (no need to set a hacky name-to-rank map).

Also, will these tests be push-blocking? A lot of these tests will fail due to unimplemented functionality, so just want to ensure that PR's don't get reverted due to these failures. Otherwise we can disable the failing tests and enable them as we implement more features.

lw committed:
Update on "[TensorPipe] Do not require user to provide worker name-to-rank map"


The initial version of the TensorPipe agent required the user to specify the full map between workers' names and their ids, on each worker. However it's enough for each worker to just specify their name and id, as these can then be exchanged using the store.

Addresses #37784, although I think we can go further and use the store to also automatically assign ranks to workers, so that the user only needs to specify a name.

Differential Revision: [D21463833](https://our.internmc.facebook.com/intern/diff/D21463833/)

[ghstack-poisoned]

lw commented:
Uhm, it looks like there has been a ghexport mishap, let me try again...

lw committed:
Update on "[TensorPipe] Do not require user to provide worker name-to-rank map"


The initial version of the TensorPipe agent required the user to specify the full map between workers' names and their ids, on each worker. However it's enough for each worker to just specify their name and id, as these can then be exchanged using the store.

Addresses #37784, although I think we can go further and use the store to also automatically assign ranks to workers, so that the user only needs to specify a name.

Differential Revision: [D21463833](https://our.internmc.facebook.com/intern/diff/D21463833/)

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Do not require user to provide worker name-to-rank map"


The initial version of the TensorPipe agent required the user to specify the full map between workers' names and their ids, on each worker. However it's enough for each worker to just specify their name and id, as these can then be exchanged using the store.

Addresses #37784, although I think we can go further and use the store to also automatically assign ranks to workers, so that the user only needs to specify a name.

Differential Revision: [D21463833](https://our.internmc.facebook.com/intern/diff/D21463833/)

[ghstack-poisoned]

lw commented:
All tests are passing, I'm landing this.

I also feel a bit queasy about the lack of tests for the TensorPipe agent. I have ran manual tests for this diff (they are in the test plan on Phabricator, not sure why they didn't get exported). I'm looking forward to enrolling this agent into the existing tests in some form, so that we can have more automatic signal...

mrshenli commented:
> Also, will these tests be push-blocking? A lot of these tests will fail due to unimplemented functionality, so just want to ensure that PR's don't get reverted due to these failures. Otherwise we can disable the failing tests and enable them as we implement more features.

Thanks @osalpekar for adding the test. I don't think our RPC tests are push-blocking, but if they consistently fail due to missing features, it's better that we only run the subset that can pass currently? Is it possible to create a black list of tests to skip for TensorPipeAgent, and gradually remove tests from that list?
